onnxruntime/contrib_ops/cpu/quantization/attention_quant.cc

```diff
@@ -91,10 +102,39 @@ Status QAttention<T>::PrePack(const Tensor& weights, int input_idx, bool& is_pac
     weights_data += head_size;
   }
 
+  if (!kernel_owns_prepacked_buffer) {
+    prepacked_weight_for_caching.buffers_.push_back(std::move(packed_weights_));
+    prepacked_weight_for_caching.shapes_.push_back(weight_shape_);
+    prepacked_weight_for_caching.weights_sizes_.push_back(packed_weights_size_);
+    prepacked_weight_for_caching.has_cached_ = true;
+    packed_weights_ = BufferUniquePtr(prepacked_weight_for_caching.buffers_[0].get(), BufferDeleter(nullptr));
+  }
+
   is_packed = true;
   return Status::OK();
 }
 
+template <typename T>
+Status QAttention<T>::UseCachedPrePackedWeight(const PrepackedWeight& cached_prepacked_weight,
+                                               int input_idx,
+                                               /*out*/ bool& read_from_cache) {
+  read_from_cache = false;
+
+  if (1 != input_idx) {
+    return Status::OK();
+  }
+
+  // Cached pre-packed weight
+  if (cached_prepacked_weight.has_cached_) {
```

Comment:
This should be checked before invoking. Kernels shouldn't be burdened with this.

Reply:
Sure- removed this